utils/man/grenum.1
.TH grenum 1 "January 15th, 2011" Version 1.6.2.20110115

.SH NAME
.B grenum
\- An advanced refdes renumber utility

.SH SYNOPSIS
.B grenum
[\-v\||\|\-\-version \-h\||\|\-\-help \-p\||\|\-\-pagejump]
.I infile1.sch infile2.sch
\|.\|.\|.


.SH DESCRIPTION

The
.B grenum
program will renumber the refdes definitions read from
.I infile\|.

.SH OPTIONS
.TP
\-p, \-\-pagejump
This will switch on the page jump mode. It means that refdes's from each
.I inputfile
gets renumbered from 100,200... etc.
.TP
\-v, \-\-version
prints version information
.TP
\-h, \-\-help
prints help message

.SH RETURN VALUE
.B grenum
will return
.B 0
if all files processed successfully,
.B 1
if there was no input file specified,
.B 2
if I/O files can't be opened/written,
.B 3
if there is parse error in input file,
.B 4
if there is not enough memory to store refdes prefixes.
.SH BUGS

If you find one, please report it to:

.SH AUTHOR

Levente Kovacs

Levente.Kovacs@interware.hu

.SH SEE ALSO

.B gschem(1)


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2011 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf 1.7.0
=================================

* All gEDA/gaf applications now handle configuration files much more
  strictly, and provide much more useful feedback when an error
  occurs.

* Several changes to `gschem' command-line options:

  - The `-t' option is no longer supported.

  - Several command-line options now support a long-form alternative.

  - The `-s' option can now be used to run multiple Scheme scripts
    during `gschem' startup.

  - A new `-c' option is available, which runs a Scheme expression
    during `gschem' startup.

  - A new `-L' option is available, which adds a directory to the
    Scheme load path.

  - A new `-V' or `--version' option is available, which displays
    version information.

* `gschem' now provides a "Select All" function, which is bound to
  <Control a> by default, along with a corresponding "Deselect"
  function.

* In the `gschem' "Add component" dialog, filtering the component
  library will expand matching libraries.  If the filter is cancelled,
  all libraries are collapsed.  Components can now be selected by
  double-clicking on them in the component list.

* The `gschem' "Edit Text Properties" dialog now has an easier to use
  "Alignment" menu.

* Printed text size now better matches on-screen text size in
  `gschem'.

* Special characters, including commas, are now escaped or quoted
  appropriately in `gattrib' CSV output.

* Several changes to `gnetlist' command-line options:

  - Four command-line options specific to the `spice-sdb' backend have
    been removed.  They can be replaced with `-O <option>' according
    to the following table:

        -e, --embedd              -O embedd_mode
        -I, --include             -O include_mode
        -n, --nomunge             -O nomunge_mode
        -s, --sort                -O sort_mode

  - The list of available backends is now obtained using a new
    `--list-backends' option.  `-g help' no longer outputs a list of
    backends.

  - A new `-L' option is available, which adds a directory to the
    Scheme load path.

  - A new `-V' or `--version' option is available, which displays
    version information.

  - When run with invalid command-line arguments, `gnetlist' now exits
    with non-zero exit status.

* Several `gnetlist' backends have now been fixed so that it should no
  longer be necessary to set an expanded Guile stack in gEDA's config
  files.

* The `spice-sdb' backend for `gnetlist' now supports probes
  (`TESTPOINT' devices).

* The undocumented `mk_verilog_syms' program is no longer installed.
  The symbols it creates are still installed to
  `${prefix}/share/gEDA/sym/verilog'.

Notable changes in gEDA/gaf 1.6.2
=================================

 * PNG export from `gschem' will now be cropped correctly.

 * The `partlist' backends for `gnetlist' now handle hierarchical
   refdes correctly.

 * On Windows, `gschem' and `gattrib' will no longer launch with a
   command window visible.

 * `gnetlist-arg' is no longer valid in `gsch2pcb' project files.

Several other bugs were fixed, including a major crash bug in
gnetlist.

Notable changes in gEDA/gaf 1.6.1
=================================

* Added accelerator mnemonics for root menus in gschem

* Lots of updated and new language translations in libgeda, gschem, and
  gattrib.  (Courtesy of language translators and Launchpad) 

* Updated the shipped documentation from the official gEDA wiki

* Fixed the following bugs:

  - 2876373: Fix rendering with zoomed out dashed lines
  - 2901183: Fix for some manpage issues
  - 2904715: Don't update page connectivity from o_place_rotate()
  - 2949232: Fix slot editing routines to not edit inherited attributes 
  - Cherry picked some other bug fixes from master that should be in
    this release

Notable changes in gEDA/gaf 1.6.0
=================================

* Fixed the GtkItemEntry code inside of gattrib code so that it builds
  using gtk+ > 2.16.x.

* In gschem, rubberband pin to pin connections by adding nets.

* Cleanup of the slot handling code in libgeda.

* Fixes/updates to the MinGW port.

* Updated the shipped documentation from the online wiki.

Notable changes in gEDA/gaf 1.5.3/1.5.4
=======================================

* NOTE: gEDA/gaf 1.5.3 was released with some critical bugs, so
        it was withdrawn before it was widely distributed.

* Text rendering using native system fonts:

  - The various programs in gEDA/gaf no longer uses its own built-in
    line font.

  - Native system fonts are used for rendering, giving support for a
    wider range of symbols, and better looking schematics.

* Always use PostScript fonts in gschem's PostScript output.  The
  "output-text", "text-output" and "output-vector-threshold" config
  functions have been removed.

* Image rendering in 'gschem' uses cairo instead of GDK.

* Numerous build system changes:

  - gEDA/gaf is now distributed as a single source archive, called
    gEDA-<version>.tar.gz.

  - GTK+ 2.10 or newer and Guile 1.8 or newer are now required.

  - `intltool' is no longer required.

  - Documentation is now installed to ${docdir} (usually
    ${prefix}/share/doc/gEDA).

* The attribute edit dialog in `gschem' now has completion of most
  common attribute names.

* The multi-attribute edit dialog in `gschem' now shows unpromoted
  symbol attributes.

* Menu items in `gschem' now have icons and properly aligned key
  bindings. They also support accelerator keys.

* The `gschem2pcb' and `PCBboard' netlist backends have been removed.
  It is recommended to use `gsch2pcb' instead.

* Check for and reject non-footprint PCB files in 'gsch2pcb'.


utils/src/grenum.c
/* $Id$ */
/*	This is grenum, an advanced refdes renumber utility for gEDA's gschem.
 *
 *	Copyright (C) 2005-2010  Levente Kovacs
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation; either version 2 of the License, or
 *	(at your option) any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with this program; if not, write to the Free Software
 *	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 *
 * Levente.Kovacs@interware.hu
 *
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "version.h"

#include <stdio.h>
#include <stdlib.h>

#ifdef HAVE_STRING_H
#include <string.h>
#endif

//#define _GNU_SOURCE

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#include <errno.h>

#include "grenum.h"
/*#define DEBUG*/

int main(int argc, char *argv[])
	{
	FILE *infile,*outfile;
	char buff[BUFFSIZE], infilename[FILENAMESIZE], outfilename[FILENAMESIZE];
	unsigned char flags;
	int c, pages, ret;
	unsigned int i;

#ifdef HAVE_GETOPT_LONG
	int opt_idx;
	const struct option long_opts[]={
		{"version",  no_argument, NULL, 'v'},
		{"help",     no_argument, NULL, 'h'},
		{"pagejump", no_argument, NULL, 'p'},
		{0,0,0,0}};
#endif
	const char *opt_options = "vhp";

	struct refdes_ refdes, refdes_db[MAX_PREFIX_COUNT];

	flags=0x00; /*Clear all flags*/

	while(1)
		{
#ifdef HAVE_GETOPT_LONG
		c = getopt_long(argc, argv, opt_options, long_opts, &opt_idx);
#else
		c = getopt(argc, argv, opt_options);
#endif
		if(c == -1)
			break;
		switch(c)
			{
			case 'h':
				printhelp();
				return 0;
			case 'v':
				printver();
				return 0;
			case 'p':
				flags|=PAGEJUMP; /*Set the pagejump flag*/
				break;
			}
		}
	if(optind==argc)
		{
		printf("grenum: no input file\n");
		printhelp();
		return NO_INPUT_FILE;
		}
	for(c=0;c<MAX_PREFIX_COUNT;++c)
		{
		refdes_db[c].prefix[0]='\0';	/*Zero all the strings in the database.*/
		refdes_db[c].value=COUNT_START;
		}

	for(pages=1;optind<argc;++optind,++pages)
		{
		if((flags&PAGEJUMP)==PAGEJUMP)	/*pagejumps*/
			{
			for(c=0;c<MAX_PREFIX_COUNT;++c)
				refdes_db[c].value=PAGE_JMP*pages+COUNT_START;	/*Reset the counters according to page numbers*/
			}
		strcpy(&infilename[0],argv[optind]);	/*Copy the filename to the buffer*/
		if((infile=fopen(infilename, "r")) == NULL)	/*Open file, use r+ for read and write*/
			{
			perror("grenum: unable to open input file");
			return FILE_OP_ERROR;
			}
		strcpy(&outfilename[0],&infilename[0]);
		if((outfile=fopen(strcat(&outfilename[0],".tmp"),"wb"))==NULL)
			{
			perror("grenum: could not create tmp file");
			fclose(infile);	/*Close the file*/
			return FILE_OP_ERROR;
			}
		printf("grenum: processing file %s\n",&infilename[0]);
		while((ret=get_refdes_from_file(infile, &refdes, buff))!=END_OF_FILE) /*Read one line.*/
			{	/*Process starts here*/
#ifdef DEBUG
			printf("%s\n",&buff[0]);	/*Print out what is read*/
#endif
			switch(ret)
				{
				case NOT_REFDES_LINE:
				if(fputs(buff,outfile)==-1)
					{
					perror("grenum: could not write to tmp file");
					fclose(infile);	/*Close the files*/
					fclose(outfile);
					return FILE_OP_ERROR;
					}
				continue;
				case REFDES_WITH_VALUE:	/*We shall compare the maximum value, shall search for gaps, and set the refes_db.value to the next  free value*/
				c=refdes_lookup(refdes_db, &refdes);
				switch(c)
					{
					case REFDES_NOT_FOUND:	/*No such prefix*/
					strcpy(&refdes_db[refdes.prefixes].prefix[0],&refdes.prefix[0]);	/*Register the prefix to the database*/
					refdes_db[refdes.prefixes+1].prefix[0]='\0';
					refdes_db[refdes.prefixes].value=refdes.value;	/*Renumber... Finally :-)*/
					break;
					case MAX_PREFIX_COUNT:	/*Out of memory*/
					printf("grenum: out of memory. Too much refdes prefixes.\n");
					fclose(infile);	/*Close the files*/
					fclose(outfile);
					return OUT_OF_MEMORY;
					default:
					if(refdes.value-refdes_db[c].value==1)	/*If we have the next value, don't do anything, just update the database.*/
						{
						refdes_db[c].value=refdes.value;
						break;
						}
				/*Now we have a hole in numbering. Let's see if it'll be fixed, and seek for the maximum value. eg. R1,R2,R5,R3. So, we have to search for R3,R4, and set the db to R3.*/

					for(i=refdes_db[c].value+1; i<refdes.value; ++i)
						{
						if(seek_value(c, infile, i, refdes_db)==VALUE_NOT_FOUND)
							{
							refdes_db[c].value=i-1;
							break;
							}
						}
					if(i!=refdes.value)
						flags|=GAP_DETECTED;
					else
						flags|=~GAP_DETECTED;
					break;
					}
				break;	/*continue our job*/
				case REFDES_WITHOUT_VALUE:
				c=refdes_lookup(refdes_db, &refdes);
				switch(c)
					{
					case -1:	/*No such prefix*/
					strcpy(&refdes_db[refdes.prefixes].prefix[0],&refdes.prefix[0]);
					refdes.value=++refdes_db[refdes.prefixes].value;
					refdes_db[refdes.prefixes+1].prefix[0]='\0';
					break;
					case MAX_PREFIX_COUNT:
					printf("grenum: out of memory. Too much refdes prefixes.\n");
					fclose(infile);	/*Close the files*/
					fclose(outfile);
					return OUT_OF_MEMORY;
					default:
					if((flags&GAP_DETECTED)==GAP_DETECTED)
						{
						for(i=refdes_db[c].value+1; seek_value(c, infile, i, refdes_db)!=VALUE_NOT_FOUND; ++i);
						refdes.value=refdes_db[c].value=i;
						}
					else
						refdes.value=++refdes_db[c].value;	/*renumber*/
					break;
					}
				sprintf(buff, "refdes=%s%d\n", &refdes.prefix[0], refdes.value);
				break;
				case REFDES_ERROR:	/*e.g. awdf#$%WSf82f8 :-) No "=" signal in the refdes string.*/
					printf("grenum: parse error\n");
					fclose(infile);	/*Close the files*/
					fclose(outfile);
					return PARSE_ERROR;
				}
			if(fputs(buff,outfile)==-1)	/*Finally, write the refdes line to the output file*/
				{
				perror("grenum: could not write to tmp file");
				fclose(infile);	/*Close the files*/
				fclose(outfile);
				return FILE_OP_ERROR;
				}
			}	/*Process ends here*/

		fclose(outfile);
		strcpy(&buff[0],&infilename[0]); /*buff has the original infilename*/
/*The next few lines implements the copy program*/
		fseek(infile,0L,SEEK_SET); /*Go to the begining of the infile*/
		outfile=fopen(strcat(&buff[0],".save"),"wb");
		if(outfile==NULL)
			{
			perror("grenum: ould not create backup file");
			fclose(infile);	/*Close the file*/
			return FILE_OP_ERROR;
			}
		while(fgets(&buff[0],BUFFSIZE,infile)!=NULL) /*Read one line.*/
			{
			if(fputs(&buff[0],outfile)==-1)
				{
				perror("grenum: could not write to backup file");
				fclose(infile);	/*Close the files*/
				fclose(outfile);
				return FILE_OP_ERROR;
				}
			}
		fclose(infile);
		fclose(outfile);
		rename(outfilename, infilename);	/*Move the tmpfile to the original*/
		}
	printf("grenum: file(s) successfully processed\n");
	return OK; /*Everything is okay*/
	}

int get_refdes_from_file(FILE *fp, struct refdes_ *refdes, char *buff)
	{

/*	Read one line from file, and return the following things:
 *
 *	END_OF_FILE if file reaches its end. The content of buff is unknown!
 *
 * NOT_REFDES_LINE if the current line is not a refdes. The line will
 * saved in buff.
 *
 * Return according to parse_refdes(). The buff will contain the current line too.
 */

	if(fgets(buff, BUFFSIZE, fp)==NULL)
		return END_OF_FILE;
	if(strstr(buff, "refdes=")==NULL)
		return NOT_REFDES_LINE;
	return parse_refdes(refdes, buff);
	}

int seek_value(int prefix, FILE *fp, unsigned int value, struct refdes_ *db)
	{
	fpos_t filepos;
	int ret;
	struct refdes_ refdes;
	char buff[BUFFSIZE];

	fgetpos(fp, &filepos);	/*First of all, save the file pos.*/
	rewind(fp);	/*Rewind*/
	while((ret=get_refdes_from_file(fp, &refdes, buff))!=END_OF_FILE)
		{
		if(ret==REFDES_WITH_VALUE && prefix==refdes_lookup(db, &refdes) && refdes.value==value)
			{
			fsetpos(fp,&filepos);
			return VALUE_FOUND;
			}
		}
	fsetpos(fp,&filepos);
	return VALUE_NOT_FOUND;
	}

int parse_refdes(struct refdes_ *refdes, char *ref_str)
	{
	int i;
	char buff[BUFFSIZE],*cpr,*cp;

/*
 * This function parses the refdes line from the .sch file. It takes a pointer to the
 * complete refdes definition string, and a pointer which points to a refdes structure
 * where it'll store the info.
 *
 * parse_refdes() will return
 *
 * REFDES_WITH_VALUE if there was a prefix with renumbered value
 * (for example R1,IC3,U5);
 *
 * REFDES_WITHOUT_VALUE if there was a "?" mark found, and it has to be
 * renumbered (e.g. U?);
 *
 * REFDES_ERROR, if there was some uncool thing.
 *
 * The refdes structure is filled with the prefix and the value.
 *
 * Note that if a "?" is found, the value member remains untouched.
 */


	cpr=strstr(ref_str,"=");	/*seek for the "="*/
	if(cpr==NULL)	/*This should not happen*/
		return REFDES_ERROR;
	cp=strstr(ref_str,"?");
/*refdes=U1		refdes=IC?
 *      |		         |
 *    *cpr		        cp            
 */
	if(cp!=NULL)
		{	/*Not renumbered yet*/
		strncpy(&refdes->prefix[0],cpr+1,cp-cpr-1);	/*Copy the prefix to the refdes structure*/
		refdes->prefix[cp-cpr-1]='\0';

#ifdef DEBUG
		printf("Prefix=%s\n",&refdes->prefix[0]);
#endif

		return REFDES_WITHOUT_VALUE;
		}
	for(cp=cpr+1,i=0;(*cp != '\n' && *cp>='A' && *cp<='z');++i,++cp) /*No "?". Copy the prefix*/
		buff[i]=*cp;	/*Fill the buffer from char to char*/
	buff[i]='\0';	/*Terminate with NULL to be a string*/
#ifdef DEBUG
	printf("Prefix=%s\n",&buff[0]);
#endif
	strcpy(&refdes->prefix[0],&buff[0]);	/*Copy to refdes structure*/
	for(i=0,cp;(*cp != '\n' && *cp>='0' && *cp<='9');++cp,++i)
		buff[i]=*cp;	/*Fill the buffer from char to char*/
	buff[i]='\0';	/*Terminate with NULL to be a string*/
#ifdef DEBUG
	printf("Value=%s\n",&buff[0]);
#endif
	refdes->value=abs(atoi(&buff[0]));
	return REFDES_WITH_VALUE;
	}

int refdes_lookup(struct refdes_ *db, struct refdes_ *ref)
	{
	int c=0;

	for(c=0;c<MAX_PREFIX_COUNT;++c)
		{
		if(strcmp(ref->prefix,(*(db+c)).prefix)==0)
			break;
		else if((*(db+c)).prefix[0]=='\0')
			{
			ref->prefixes=c;
			return REFDES_NOT_FOUND;
			}
		}
	return c;
	}

void printhelp()
	{
#ifdef HAVE_GETOPT_LONG
	const char *v_opt="-v | --version";
	const char *h_opt="-h | --help";
	const char *p_opt="-p | --pagejump";
#else
	const char *v_opt="-v";
	const char *h_opt="-h";
	const char *p_opt="-p";
#endif

	printver();
	printf("Usage: grenum [%s] [%s] [%s] file1.sch file2.sch ...\n\n",
		v_opt, h_opt, p_opt);
	printf("\t%s\tprints version info\n\t%s\tprints this help\n\t%s\tsets pagejump mode on\n",
		v_opt, h_opt, p_opt);
	printf("For more information read the README file and/or the manual.\n");
	}


void printver()
	{
	printf("This is grenum, an advanced refdes renumber utility for gEDA's gschem.\n");
	printf("Version %s.  gEDA/gaf version %s.%s\n",GRVERSION, 
               PACKAGE_DOTTED_VERSION, PACKAGE_DATE_VERSION);
	printf("Compiled on %s at %s\n",COMP_DATE,COMP_TIME);
	}
